core/deployment/src/main/java/io/quarkus/deployment/dev/filesystem/ReloadableFileManager.java

```diff
@@ -0,0 +1,379 @@
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }, this.reloadableFileManager == null ? null : () -> {
+            try {
+                return this.reloadableFileManager.listLocationsForModules(location);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+    }
+
+    private <T> Iterable<T> combinedIterable(Supplier<Iterable<T>> required, Supplier<Iterable<T>> optional) {
+        if (optional == null) {
+            return required.get();
+        }
+        final Iterable<T> requiredIterable = required.get();
+        if (requiredIterable == null) {
+            return optional.get();
+        }
+        final Iterable<T> optionalIterable = optional.get();
+        if (optionalIterable == null) {
+            return requiredIterable;
+        }
+        return () -> {
+            final Iterator<T> requiredI = requiredIterable.iterator();
+            final Iterator<T> optionalI = optionalIterable.iterator();
+            if (!requiredI.hasNext()) {
+                return optionalI;
+            }
+            if (!optionalI.hasNext()) {
+                return requiredI;
+            }
+            return new Iterator<>() {
+
+                boolean currentIsStatic = true;
+
+                @Override
+                public boolean hasNext() {
+                    return requiredI.hasNext() || optionalI.hasNext();
+                }
+
+                @Override
+                public T next() {
+                    currentIsStatic = requiredI.hasNext();
+                    if (currentIsStatic) {
+                        return requiredI.next();
+                    }
+                    return optionalI.next();
+                }
+
+                @Override
+                public void remove() {
+                    if (currentIsStatic) {
+                        requiredI.remove();
+                    } else {
+                        optionalI.remove();
+                    }
+                }
+            };
+        };
+    }
+
+    @Override
+    public boolean contains(Location location, FileObject fo) throws IOException {
+        return super.contains(location, fo) || (this.reloadableFileManager != null
+                && this.reloadableFileManager.contains(location, fo));
+    }
+
+    @Override
+    public Iterable<? extends JavaFileObject> getJavaSources(Iterable<? extends File> files) {
+        return this.fileManager.getJavaFileObjectsFromFiles(files);
+    }
+
+    @Override
+    public void reset(Context context) {
+        try {
+            this.reloadableFileManager.close();
+            this.reloadableFileManager.setLocation(StandardLocation.CLASS_PATH, context.getReloadableClassPath());
+        } catch (IOException e) {
+            throw new RuntimeException("Cannot reset reloadable file manager", e);
+        }
+        super.reset(context);
+    }
```

Comment:
First, close and reset the reloadable file manager, then forward the static file manager reset to super class